quint/src/types/constraintGenerator.ts

```diff
@@ -38,14 +38,39 @@ import { getSignatures } from './builtinSignatures'
 import { Constraint, Signature, TypeScheme, toScheme } from './base'
 import { Substitutions, applySubstitution, compose } from './substitutions'
 import { LookupTable } from '../names/base'
-import { specialConstraints } from './specialConstraints'
+import {
+  fieldConstraints,
+  fieldNamesConstraints,
+  itemConstraints,
+  recordConstructorConstraints,
+  tupleConstructorConstraints,
+  withConstraints,
+} from './specialConstraints'
 import { FreshVarGenerator } from '../FreshVarGenerator'
 
 export type SolvingFunctionType = (
   _table: LookupTable,
   _constraint: Constraint
 ) => Either<Map<bigint, ErrorTree>, Substitutions>
 
+function validateArity(
+  operator: string,
+  args: [QuintEx, QuintType][],
+  pred: (arity: number) => Boolean,
+  msg: String
+): Either<Error, null> {
+  if (!pred(args.length)) {
+    return left(
+      buildErrorLeaf(
+        `Checking arity for application of ${operator}`,
```

Comment:
The first argument to `buildErrorLeaf` is expected to be a location.
Should we move this inside `ConstraintGeneratorVisitor` so we can use its `location` field?

Reply:
That's not really necessary. This location is the `leaf` of the error tree. The returned error here will be combined with the `ConstraintGeneratorVisitor`'s location in its `addToResults` method, and the resulting error will look somewhat like:
```
Error: Operator expects 2 arguments but given 1
Checking arity for operator application of foo
Generating constraints for foo("bar")
```

Reply:
Okay, but should we then update the docs on `buildErrorLeaf`?
It looks like `location` is more of an explanation.

Reply:
So I understand the location here to be the context in which the thing is happening. That is inferred from the use I’ve seen (e.g. in the special constraints). But if it’s causing confusion for Thomas maybe a rename of the param or added docs is a good idea. I think it’s out of scope for this PR tho, iiuc, and can be addressed in a followup.